aten/src/ATen/native/cuda/MultinomialKernel.cu

```diff
@@ -2,13 +2,506 @@
+      // All threads in bounds load a value
+      int cat = chunk * blockDim.x + threadIdx.x;
+
+      accscalar_t a_dist_val = cat < categories ?
+                               static_cast<accscalar_t>(dist[curDist * stride_dist + cat * stride_categories]) / sum :
+                               accZero;
+      scalar_t dist_val = static_cast<scalar_t>(a_dist_val);
+
+      smem[threadIdx.x] = dist_val;
+      __syncthreads();
+
+      // Perform an inclusive prefix sum of the shared memory contents
+      for (int offset = 1; offset < blockDim.x; offset *= 2) {
+        scalar_t val = zero;
+
+        if (threadIdx.x >= offset) {
+          val = smem[threadIdx.x - offset] + smem[threadIdx.x];
+        }
+
+        __syncthreads();
+        if (threadIdx.x >= offset) {
+          smem[threadIdx.x] = val;
+        }
+        __syncthreads();
+      }
+
+      // Each thread will check to see if the sample falls in its
+      // bucket
+      scalar_t curBucket = smem[threadIdx.x] + prevHighProb;
+      scalar_t prevBucket =
+          threadIdx.x == 0 ? prevHighProb :
+          smem[threadIdx.x - 1] + prevHighProb;
+      bool inBucket =
+          (cat < categories) &&
+          (!(sample >= curBucket) &&
+           (sample >= prevBucket) &&
+           (dist_val > zero));
+
+      if (inBucket) {
+        // We're done; we have the sample
+        // Torch indices are 1-based
+        dest[curDist] = cat;
+        found = true;
+      }
+
+      // Store the previous scan's high value for future use
+      prevHighProb = prevHighProb + smem[blockDim.x - 1];
+
+      __syncthreads();
+    }
+
+    if (threadIdx.x == 0 && !found) {
+      // This should address a rare bug where we don't select a valid index. This likely occurs when
+      // due to floating point arithmetic rounding errors, our cumulative sum does not add up to 1, but
+      // and our uniform sample is greater than this value. In this case we likely have unitialized memory
+      // in dest[curDist]. So basically we will loop through the distribution and pick the largest index
+      // where the distribution is non-zero. This is obviously terribly inefficient, but due to the
+      // rarity in which this occurs, this should not be an issue.
+      for (int cat = categories - 1; cat >= 0; --cat) {
+        if (dist[curDist * stride_dist + cat * stride_categories] > zero) {
+          dest[curDist] = cat;
+          break;
+        }
+      }
+    }
+  }
+}
+
+
+void multinomial_kernel_impl(Tensor& result, const Tensor& self, const int64_t n_sample, const bool with_replacement, Generator* generator) {
+  auto gen = get_generator_or_default<CUDAGenerator>(generator, cuda::detail::getDefaultCUDAGenerator());
+
+  int inputSize = self.dim();
+  int64_t numDist =
+      inputSize == 1 ? 1 : self.size(0);
+  int numCategories =
+      inputSize == 1 ? self.size(0) : self.size(1);
+
+  // Restructure data for 2d
+  if (inputSize == 1) {
+    self.unsqueeze_(0);
```

Comment:
This is bad pattern (to modify inputs inplace), much better is to create another view and operate with it.

Reply:
Not only is this a bad pattern, this is also not okay because Tensor should not be modified (the argument is `const Tensor&`).